File: src/core/Stocks/PositionInstance.cs
using System;
using System.Collections.Generic;
using System.Linq;
using core.Shared;

namespace core.Stocks
{
    public class PositionEventType
    {
        public const string Buy = "buy";
        public const string Stop = "stop";
        public const string Sell = "sell";
        public const string Risk = "risk";

        public PositionEventType(string value)
        {
            Value = value switch
            {
                Buy => Buy,
                Stop => Stop,
                Sell => Sell,
                Risk => Risk,
                _ => throw new InvalidOperationException($"Invalid position event type: {value}")
            };
        }

        public string Value
        {
            get; 
        }
    }
    
    public readonly record struct PositionEvent (Guid Id, string Description, PositionEventType Type, decimal? Value, DateOnly When, decimal? Quantity = null)
    {
        public readonly string Date => When.ToString("yyyy-MM-dd");
    }
    public readonly record struct PositionTransaction(decimal NumberOfShares, decimal Price, Guid TransactionId, string Type, DateTimeOffset When)
    {
        public readonly string Date => When.ToString("yyyy-MM-dd");
        public readonly int AgeInDays => (int)(DateTimeOffset.UtcNow - When).TotalDays;
    }

    // PositionInstance models a stock position from the time the first share is opened
    // to the time when the last share is sold.
    public class PositionInstance
    {
        public PositionInstance(int positionId, Ticker ticker, DateTimeOffset opened)
        {
            PositionId = positionId;
            Ticker = ticker;
            Opened = opened;
        }

        public decimal NumberOfShares { get; private set; }
        public decimal AverageCostPerShare { get; private set; }
        public decimal AverageSaleCostPerShare { get; private set; }
        public decimal AverageBuyCostPerShare { get; private set; }
        public DateTimeOffset Opened { get; }
        public int DaysHeld => (int)(Closed ?? DateTimeOffset.UtcNow).Subtract(Opened).TotalDays;
        public decimal Cost { get; private set; }
        public decimal Profit { get; private set; }
        public decimal GainPct => IsClosed switch {
            true => (AverageSaleCostPerShare - AverageBuyCostPerShare) / AverageBuyCostPerShare,
            false => 0
        };
            
        public decimal RR => RiskedAmount switch {
            not null => 
                RiskedAmount.Value switch
                {
                    0 => 0,
                    _ => Profit / RiskedAmount.Value
                },
            _ => 0
        };
        public decimal RRWeighted => RR * Cost;

        public bool IsClosed => Closed != null;
        public int PositionId { get; }
        public Ticker Ticker { get; }
        public DateTimeOffset? Closed { get; private set; }
        public decimal? FirstStop { get; private set; }
        public decimal? RiskedAmount { get; private set; }
        public decimal? CostAtRiskedBasedOnStopPrice => StopPrice switch {
            not null => StopPrice > AverageCostPerShare ? 0 : (AverageCostPerShare - StopPrice) * NumberOfShares,
            _ => null
        };
        
        public List<PositionTransaction> Transactions { get; } = new();
        public List<PositionEvent> Events { get; } = new();

        public decimal? StopPrice { get; private set; }
        public DateTimeOffset LastTransaction { get; private set; }
        public decimal LastSellPrice { get; private set; }
        public int DaysSinceLastTransaction => (int)(DateTimeOffset.UtcNow - LastTransaction).TotalDays;
        private readonly List<decimal> _slots = new();

        private bool _positionCompleted;
        public decimal CompletedPositionCost { get; private set; }
        public decimal CompletedPositionShares { get; private set; }
        public decimal CompletedPositionCostPerShare => CompletedPositionCost / CompletedPositionShares;

        public TradeGrade? Grade { get; private set; }
        public string GradeNote { get; private set; }
        public void SetGrade(TradeGrade grade, string note = null)
        {
            Grade = grade;
            var previousNote = GradeNote;
            GradeNote = note;
            Notes.Remove(previousNote);
            Notes.Add(note);
        }

        public List<string> Notes { get; } = new();
        public void AddNotes(string notes)
        {
            Notes.Add(notes);
        }

        public void Buy(decimal numberOfShares, decimal price, DateTimeOffset when, Guid transactionId, string notes = null)
        {
            Transactions.Add(new PositionTransaction(numberOfShares, price, TransactionId: transactionId, Type:"buy", when));
            Events.Add(new PositionEvent(transactionId, $"buy {numberOfShares} @ ${price}", Type: new PositionEventType(PositionEventType.Buy), Value: price, When: DateOnly.FromDateTime(when.DateTime), Quantity: numberOfShares));

            if (_positionCompleted == false)
            {
                CompletedPositionCost += price * numberOfShares;
                CompletedPositionShares += numberOfShares;
            }

            if (notes != null)
            {
                Notes.Add(notes);
            }

            LastTransaction = when;

            RunCalculations();
        }

        internal void RemoveTransaction(Guid transactionId)
        {
            var tx = Transactions.SingleOrDefault(t => t.TransactionId == transactionId);
            if (tx.TransactionId == default)
            {
                throw new InvalidOperationException($"Transaction {transactionId} not found");
            }
            Transactions.Remove(tx);

            var ev = Events.SingleOrDefault(e => e.Id == transactionId);
            if (ev.Id == default)
            {
                throw new InvalidOperationException($"Event {transactionId} not found");
            }
            Events.Remove(ev);
            
            RunCalculations();
        }

        public void Sell(decimal numberOfShares, decimal price, Guid transactionId, DateTimeOffset when, string notes = null)
        {
            if (NumberOfShares - numberOfShares < 0)
            {
                var details = $"Sell {numberOfShares} @ ${price} on {when} for {Ticker}";
                throw new InvalidOperationException("Transaction would make amount owned invalid: " + details);
            }

            // once we stop adding and do our first sale, that should be our position
            // completion
            if (!_positionCompleted)
            {
                _positionCompleted = true;
            }

            var percentGainAtSale = (price - AverageCostPerShare) / AverageCostPerShare;

            Transactions.Add(new PositionTransaction(numberOfShares, price, TransactionId:transactionId, Type: "sell", when));
            Events.Add(new PositionEvent(transactionId, $"sell {numberOfShares} @ ${price} ({percentGainAtSale:P})", new PositionEventType(PositionEventType.Sell), price, DateOnly.FromDateTime(when.DateTime), Quantity: numberOfShares));

            // if we haven't set the risked amount, when we set it at 5% from the first buy price?
            if (StopPrice == null && !HasEventWithDescription("Stop price deleted"))
            {
                SetStopPrice(CompletedPositionCostPerShare * 0.95m, when);
            }            

            if (notes != null)
            {
                Notes.Add(notes);
            }

            LastTransaction = when;
            LastSellPrice = price;

            RunCalculations();

            if (NumberOfShares == 0)
            {
                Closed = when;
            }
        }

        private bool HasEventWithDescription(string testDescription) => Events.Any(e => e.Description == testDescription);

        public void SetStopPrice(decimal? stopPrice, DateTimeOffset when)
        {
            if (stopPrice != null && stopPrice != StopPrice)
            {
                StopPrice = stopPrice;

                FirstStop ??= stopPrice;

                var stopPercentage = (stopPrice.Value - AverageCostPerShare) / AverageCostPerShare;

                Events.Add(
                    new PositionEvent(Guid.Empty, $"Stop price set to {stopPrice.Value:0.##} ({stopPercentage:P1})", new PositionEventType(PositionEventType.Stop), stopPrice, DateOnly.FromDateTime(when.DateTime)));

                if (RiskedAmount == null)
                {
                    SetRiskAmount((AverageCostPerShare - stopPrice.Value) * NumberOfShares, when);
                }
            }
        }

        public void DeleteStopPrice(DateTimeOffset when)
        {
            StopPrice = null;
            RiskedAmount = null;
            
            Events.Add(new PositionEvent(Guid.Empty, "Stop price deleted", new PositionEventType(PositionEventType.Stop), null, DateOnly.FromDateTime(when.DateTime)));
        }

        public void SetRiskAmount(decimal riskAmount, DateTimeOffset when)
        {
            if (riskAmount == 0)
            {
                return;
            }
            
            RiskedAmount = riskAmount;

            Events.Add(new PositionEvent(Guid.Empty, $"Set risk amount to {RiskedAmount.Value:0.##}", new PositionEventType(PositionEventType.Risk), riskAmount, DateOnly.FromDateTime(when.DateTime)));
        }

        private void RunCalculations()
        {
            _slots.Clear();
            decimal cost = 0;
            decimal profit = 0;
            decimal numberOfShares = 0;
            
            decimal totalSale = 0;
            decimal totalNumberOfSharesSold = 0;

            decimal totalBuy = 0;
            decimal totalNumberOfSharesBought = 0;

            Transactions.ForEach(transaction => {
            if (transaction.Type == "buy")
            {
                for (var i = 0; i < transaction.NumberOfShares; i++)
                {
                    _slots.Add(transaction.Price);
                    cost += transaction.Price;
                    numberOfShares++;
                }

                totalBuy += transaction.Price * transaction.NumberOfShares;
                totalNumberOfSharesBought += transaction.NumberOfShares;
            }
            else
            {
                // remove quantity number of slots from the beginning of an array
                var removed = _slots.Take((int)transaction.NumberOfShares).ToList();
                _slots.RemoveRange(0, (int)transaction.NumberOfShares);
                removed.ForEach(
                    removedElement =>
                    {
                        profit += transaction.Price - removedElement;
                        cost -= removedElement;
                        numberOfShares--;
                    }
                );
                
                totalSale += transaction.Price * transaction.NumberOfShares;
                totalNumberOfSharesSold += transaction.NumberOfShares;
            }
            });

            // calculate average cost per share using slots
            AverageCostPerShare = _slots.Count switch {
                0 => 0,
                _ => _slots.Sum() / _slots.Count
            };
            Cost = cost;
            Profit = profit;
            NumberOfShares = numberOfShares;

            AverageSaleCostPerShare = totalNumberOfSharesSold switch {
                0 => 0,
                _ => totalSale / totalNumberOfSharesSold
            };

            AverageBuyCostPerShare = totalNumberOfSharesBought switch {
                0 => 0,
                _ => totalBuy / totalNumberOfSharesBought
            };
        }

        private readonly Dictionary<string, string> _labels = new();
        public IEnumerable<KeyValuePair<string, string>> Labels => _labels;
        public bool ContainsLabel(string key, string value)
        {
            if (!ContainsLabel(key))
            {
                return false;
            }

            return _labels[key] == value;
        }

        internal void SetLabel(PositionLabelSet labelSet)
        {
            _labels[labelSet.Key] = labelSet.Value;
        }

        public bool ContainsLabel(string key)
        {
            return _labels.ContainsKey(key);
        }

        internal void DeleteLabel(PositionLabelDeleted labelDeleted)
        {
            _labels.Remove(labelDeleted.Key);
        }

        public string GetLabelValue(string key) => _labels[key];
    }
}

File: src/core2.fs/Stocks/PositionInstance.fs
namespace core.Stocks

open System
open System.Collections.Generic
open System.Linq
open core.Shared

type PositionEventType(value: string) =
    static member Buy = "buy"
    static member Stop = "stop"
    static member Sell = "sell"
    static member Risk = "risk"
    static member Label = "label"

    member _.Value =
        match value with
        | "buy" -> PositionEventType.Buy
        | "stop" -> PositionEventType.Stop
        | "sell" -> PositionEventType.Sell
        | "risk" -> PositionEventType.Risk
        | "label" -> PositionEventType.Label
        | _ -> raise (InvalidOperationException(sprintf "Invalid position event type: %s" value))

[<Struct>]
type PositionEvent =
    { Id: Guid
      Description: string
      Type: PositionEventType
      Value: decimal option
      When: DateTimeOffset
      Quantity: decimal option
      Notes: string }
    member this.Date = this.When.ToString("yyyy-MM-dd")

[<Struct>]
type PositionTransaction =
    { NumberOfShares: decimal
      Price: decimal
      TransactionId: Guid
      Type: string
      When: DateTimeOffset }
    member this.Date = this.When.ToString("yyyy-MM-dd")
    member this.AgeInDays = int ((DateTimeOffset.UtcNow - this.When).TotalDays)

[<AllowNullLiteral>]
type PositionInstance(positionId: int, ticker: Ticker, opened: DateTimeOffset) =
    let mutable _numberOfShares = 0m
    let mutable _averageCostPerShare = 0m
    let mutable _averageSaleCostPerShare = 0m
    let mutable _averageBuyCostPerShare = 0m
    let mutable _cost = 0m
    let mutable _profit = 0m
    let mutable _closed: DateTimeOffset option = None
    let mutable _firstStop: decimal option = None
    let mutable _riskedAmount: decimal option = None
    let mutable _stopPrice: decimal option = None
    let mutable _lastTransaction = DateTimeOffset.MinValue
    let mutable _lastSellPrice = 0m
    let mutable _positionCompleted = false
    let mutable _completedPositionCost = 0m
    let mutable _completedPositionShares = 0m
    let mutable _grade: TradeGrade option = None
    let mutable _gradeDate: DateTimeOffset option = None
    let mutable _gradeNote: string = null
    let _slots = List<decimal>()
    let _transactions = List<PositionTransaction>()
    let _events = List<PositionEvent>()
    let _notes = List<string>()
    let _labels = Dictionary<string, string>()

    member _.NumberOfShares = _numberOfShares
    member _.AverageCostPerShare = _averageCostPerShare
    member _.AverageSaleCostPerShare = _averageSaleCostPerShare
    member _.AverageBuyCostPerShare = _averageBuyCostPerShare
    member _.Opened = opened
    member this.DaysHeld = 
        int ((match _closed with Some c -> c | None -> DateTimeOffset.UtcNow).Subtract(opened).TotalDays)
    member _.Cost = _cost
    member _.Profit = _profit
    member this.GainPct =
        match _averageSaleCostPerShare with
        | 0m -> 0m
        | _ -> (_averageSaleCostPerShare - _averageBuyCostPerShare) / _averageBuyCostPerShare
    member this.RR =
        match _riskedAmount with
        | Some ra when ra <> 0m -> _profit / ra
        | _ -> 0m
    member this.RRWeighted = this.RR * _cost
    member _.IsClosed = _closed.IsSome
    member _.PositionId = positionId
    member _.Ticker = ticker
    member _.Closed = _closed
    member _.FirstStop = _firstStop
    member _.RiskedAmount = _riskedAmount
    member this.CostAtRiskedBasedOnStopPrice =
        match _stopPrice with
        | Some sp when sp > _averageCostPerShare -> Some 0m
        | Some sp -> Some ((_averageCostPerShare - sp) * _numberOfShares)
        | None -> None
    member _.Transactions = _transactions :> IList<_>
    member _.Events = _events :> IList<_>
    member _.StopPrice = _stopPrice
    member _.LastTransaction = _lastTransaction
    member _.LastSellPrice = _lastSellPrice
    member this.DaysSinceLastTransaction = int ((DateTimeOffset.UtcNow - _lastTransaction).TotalDays)
    member _.CompletedPositionCost = _completedPositionCost
    member _.CompletedPositionShares = _completedPositionShares
    member this.CompletedPositionCostPerShare = _completedPositionCost / _completedPositionShares
    member _.Grade = _grade
    member _.GradeDate = _gradeDate
    member _.GradeNote = _gradeNote
    member _.Notes = _notes :> IList<_>
    member _.Labels = _labels :> IEnumerable<KeyValuePair<string, string>>

    member this.SetGrade(grade: TradeGrade, ``when``: DateTimeOffset, ?note: string) =
        _grade <- Some grade
        let previousNote = _gradeNote
        _gradeNote <- defaultArg note null
        _gradeDate <- Some ``when``
        _notes.Remove(previousNote) |> ignore
        if note.IsSome then _notes.Add(note.Value)

    member _.AddNotes(notes: string) =
        _notes.Add(notes)

    member private this.RunCalculations() =
        _slots.Clear()
        let mutable cost = 0m
        let mutable profit = 0m
        let mutable numberOfShares = 0m
        let mutable totalSale = 0m
        let mutable totalNumberOfSharesSold = 0m
        let mutable totalBuy = 0m
        let mutable totalNumberOfSharesBought = 0m

        for transaction in _transactions do
            if transaction.Type = "buy" then
                for i in 0 .. int transaction.NumberOfShares - 1 do
                    _slots.Add(transaction.Price)
                    cost <- cost + transaction.Price
                    numberOfShares <- numberOfShares + 1m
                totalBuy <- totalBuy + transaction.Price * transaction.NumberOfShares
                totalNumberOfSharesBought <- totalNumberOfSharesBought + transaction.NumberOfShares
            else
                let removed = _slots.Take(int transaction.NumberOfShares).ToList()
                _slots.RemoveRange(0, int transaction.NumberOfShares)
                for removedElement in removed do
                    profit <- profit + transaction.Price - removedElement
                    cost <- cost - removedElement
                    numberOfShares <- numberOfShares - 1m
                totalSale <- totalSale + transaction.Price * transaction.NumberOfShares
                totalNumberOfSharesSold <- totalNumberOfSharesSold + transaction.NumberOfShares

        _averageCostPerShare <-
            match _slots.Count with
            | 0 -> 0m
            | _ -> _slots.Sum() / decimal _slots.Count
        _cost <- cost
        _profit <- profit
        _numberOfShares <- numberOfShares
        _averageSaleCostPerShare <-
            match totalNumberOfSharesSold with
            | 0m -> 0m
            | _ -> totalSale / totalNumberOfSharesSold
        _averageBuyCostPerShare <-
            match totalNumberOfSharesBought with
            | 0m -> 0m
            | _ -> totalBuy / totalNumberOfSharesBought

    member this.Buy(numberOfShares: decimal, price: decimal, ``when``: DateTimeOffset, transactionId: Guid, ?notes: string) =
        _transactions.Add({ NumberOfShares = numberOfShares; Price = price; TransactionId = transactionId; Type = "buy"; When = ``when`` })
        _events.Add({ Id = transactionId; Description = sprintf "buy %M @ $%M" numberOfShares price; Type = PositionEventType(PositionEventType.Buy); Value = Some price; When = ``when``; Quantity = Some numberOfShares; Notes = defaultArg notes null })

        if not _positionCompleted then
            _completedPositionCost <- _completedPositionCost + price * numberOfShares
            _completedPositionShares <- _completedPositionShares + numberOfShares

        if notes.IsSome then _notes.Add(notes.Value)
        _lastTransaction <- ``when``
        this.RunCalculations()

    member internal this.RemoveTransaction(transactionId: Guid) =
        let tx = _transactions.SingleOrDefault(fun t -> t.TransactionId = transactionId)
        if tx.TransactionId = Guid.Empty then
            raise (InvalidOperationException(sprintf "Transaction %A not found" transactionId))
        _transactions.Remove(tx) |> ignore

        let ev = _events.SingleOrDefault(fun e -> e.Id = transactionId)
        if ev.Id = Guid.Empty then
            raise (InvalidOperationException(sprintf "Event %A not found" transactionId))
        _events.Remove(ev) |> ignore

        this.RunCalculations()

    member private this.HasEventWithDescription(testDescription: string) =
        _events.Any(fun e -> e.Description = testDescription)

    member this.Sell(numberOfShares: decimal, price: decimal, transactionId: Guid, ``when``: DateTimeOffset, ?notes: string) =
        if _numberOfShares - numberOfShares < 0m then
            let details = sprintf "Sell %M @ $%M on %O for %O" numberOfShares price ``when`` ticker
            raise (InvalidOperationException("Transaction would make amount owned invalid: " + details))

        if not _positionCompleted then
            _positionCompleted <- true

        let percentGainAtSale = (price - _averageCostPerShare) / _averageCostPerShare
        _transactions.Add({ NumberOfShares = numberOfShares; Price = price; TransactionId = transactionId; Type = "sell"; When = ``when`` })
        _events.Add({ Id = transactionId; Description = sprintf "sell %M @ $%M (%.2f%%)" numberOfShares price (float (percentGainAtSale * 100m)); Type = PositionEventType(PositionEventType.Sell); Value = Some price; When = ``when``.DateTime; Quantity = Some numberOfShares; Notes = defaultArg notes null })

        if _stopPrice.IsNone && not (this.HasEventWithDescription("Stop price deleted")) then
            this.SetStopPrice(Some (this.CompletedPositionCostPerShare * 0.95m), ``when``)

        if notes.IsSome then _notes.Add(notes.Value)
        _lastTransaction <- ``when``
        _lastSellPrice <- price
        this.RunCalculations()

        if _numberOfShares = 0m then
            _closed <- Some ``when``

    member this.SetStopPrice(stopPrice: decimal option, ``when``: DateTimeOffset) =
        match stopPrice with
        | Some sp when Some sp <> _stopPrice ->
            _stopPrice <- Some sp
            if _firstStop.IsNone then _firstStop <- Some sp
            let stopPercentage = (sp - _averageCostPerShare) / _averageCostPerShare
            _events.Add({ Id = Guid.Empty; Description = sprintf "Stop price set to %.2f (%.1f%%)" sp (stopPercentage * 100m); Type = PositionEventType(PositionEventType.Stop); Value = Some sp; When = ``when``; Quantity = None; Notes = null })
            if _riskedAmount.IsNone then
                this.SetRiskAmount((_averageCostPerShare - sp) * _numberOfShares, ``when``)
        | _ -> ()

    member this.DeleteStopPrice(``when``: DateTimeOffset) =
        _stopPrice <- None
        _riskedAmount <- None
        _events.Add({ Id = Guid.Empty; Description = "Stop price deleted"; Type = PositionEventType(PositionEventType.Stop); Value = None; When = ``when``; Quantity = None; Notes = null })

    member this.SetRiskAmount(riskAmount: decimal, ``when``: DateTimeOffset) =
        if riskAmount <> 0m then
            _riskedAmount <- Some riskAmount
            _events.Add({ Id = Guid.Empty; Description = sprintf "Set risk amount to %.2f" riskAmount; Type = PositionEventType(PositionEventType.Risk); Value = Some riskAmount; When = ``when``; Quantity = None; Notes = null })

    member this.ContainsLabel(key: string, value: string) =
        match _labels.TryGetValue(key) with
        | true, v when v = value -> true
        | _ -> false

    member internal this.SetLabel(labelSet: PositionLabelSet) =
        _labels.[labelSet.Key] <- labelSet.Value
        _events.Add({ Id = Guid.NewGuid(); Description = labelSet.Key; Type = PositionEventType(PositionEventType.Label); Value = None; When = labelSet.When; Quantity = None; Notes = labelSet.Value })

    member this.ContainsLabel(key: string) =
        _labels.ContainsKey(key)

    member internal this.DeleteLabel(labelDeleted: PositionLabelDeleted) =
        _labels.Remove(labelDeleted.Key) |> ignore
        _events.Add({ Id = Guid.NewGuid(); Description = labelDeleted.Key; Type = PositionEventType(PositionEventType.Label); Value = None; When = labelDeleted.When; Quantity = None; Notes = null })

    member this.GetLabelValue(key: string) = _labels.[key]
